src/ocaml/mode.ml

```diff
@@ -135,3 +137,135 @@ module Dict = struct
          { byte; native })
   end
 end
+
+module Select = struct
+  type mode = t
+
+  type nonrec t =
+    | Only of t
+    | All
+
+  let of_option = function
+    | None -> All
+    | Some m -> Only m
+
+  let is_not_all = function
+    | All -> false
+    | Only _ -> true
+
+  let equal t t' =
+    match (t, t') with
+    | Only m, Only m' -> equal m m'
+    | All, All -> true
+    | _, _ -> false
+
+  let hash = Poly.hash
+
+  let to_dyn t =
+    let open Dyn in
+    match t with
+    | All -> Variant ("All", [])
+    | Only m -> Variant ("Only", [ to_dyn m ])
+
+  let encode t =
+    let open Dune_sexp.Encoder in
+    match t with
+    | All -> string "all"
+    | Only m -> encode m
+
+  let decode =
+    let open Dune_sexp.Decoder in
+    let+ value = either (keyword "all") decode in
+    match value with
+    | Left () -> All
+    | Right mode -> Only mode
+end
+
+module MultiDict = struct
```

Comment:
I'm a bit uncomfortable with that data structure for the following reasons:
- as you mention this is a generalization of something that has 3 values and is unlikely to change a lot
- the "replacing" semantics of hashtbls can be unexpected when touching that piece of code
- most importantly - I don't think that this container should be mutable. we've spent a lot of energy to make most of dune's code immutable to play nice with `Memo` so I don't think we should use hashtbls in new code.

Reply:
I switched to `Map` in [7868db3](https://github.com/ocaml/dune/pull/5649/commits/7868db3e712bf26b051254af20b100dd130b980c).
It is much cleaner like this and the Map module of Dune's std even include a few helpers for multimaps in the `Multi` module that I extended a bit for our purposes.